client/pd_service_discovery.go

```diff
@@ -97,6 +100,234 @@ type ServiceDiscovery interface {
+	if c == nil || c.conn == nil {
+		return
+	}
+	healthCli := healthpb.NewHealthClient(c.conn)
+	resp, err := healthCli.Check(ctx, &healthpb.HealthCheckRequest{Service: ""})
+	failpoint.Inject("unreachableNetwork1", func() {
+		resp = nil
+		err = status.New(codes.Unavailable, "unavailable").Err()
+	})
+	rpcErr, ok := status.FromError(err)
+	if (ok && isNetworkError(rpcErr.Code())) || resp.GetStatus() != healthpb.HealthCheckResponse_SERVING {
+		c.networkFailure.Store(true)
+	} else {
+		c.networkFailure.Store(false)
+	}
+}
+
+func isNetworkError(code codes.Code) bool {
+	return code == codes.Unavailable || code == codes.DeadlineExceeded
+}
+
+// GetClientConn implements ServiceClient.
+func (c *pdServiceClient) GetClientConn() *grpc.ClientConn {
+	if c == nil {
+		return nil
+	}
+	return c.conn
+}
+
+// NeedRetry implements ServiceClient.
+func (c *pdServiceClient) NeedRetry(pdErr *pdpb.Error, err error) bool {
+	if c.IsLeader() {
+		return false
+	}
+	return !(err == nil && pdErr == nil)
+}
+
+type errFn func(pdErr *pdpb.Error) bool
+
+func regionAPIErrorFn(pdErr *pdpb.Error) bool {
+	return pdErr.GetType() == pdpb.ErrorType_REGION_NOT_FOUND
+}
+
+// pdServiceAPIClient is a specific API client for PD service.
+// It extends the pdServiceClient and adds additional fields for managing availability
+type pdServiceAPIClient struct {
+	ServiceClient
+	fn errFn
+
+	unavailable      atomic.Bool
+	unavailableUntil atomic.Value
+}
+
+func newPDServiceAPIClient(client ServiceClient, f errFn) *pdServiceAPIClient {
+	return &pdServiceAPIClient{
+		ServiceClient: client,
+		fn:            f,
+	}
+}
+
+// Available implements ServiceClient.
+func (c *pdServiceAPIClient) Available() bool {
+	return c.ServiceClient.Available() && !c.unavailable.Load()
+}
+
+func (c *pdServiceAPIClient) markAsAvailable() {
+	if !c.unavailable.Load() {
+		return
+	}
+	until := c.unavailableUntil.Load().(time.Time)
+	if time.Now().After(until) {
+		c.unavailable.Store(false)
+	}
+}
+
+// NeedRetry implements ServiceClient.
+func (c *pdServiceAPIClient) NeedRetry(pdErr *pdpb.Error, err error) bool {
+	if c.IsLeader() {
+		return false
+	}
+	if err == nil && pdErr == nil {
+		return false
+	}
+	if c.fn(pdErr) && c.unavailable.CompareAndSwap(false, true) {
+		c.unavailableUntil.Store(time.Now().Add(time.Second * 10))
+		failpoint.Inject("fastCheckAvailable", func() {
+			c.unavailableUntil.Store(time.Now().Add(time.Millisecond * 100))
+		})
+	}
+	return true
+}
+
+// pdServiceBalancerNode is a balancer node for PD service.
+// It extends the pdServiceClient and adds additional fields for the next polling client in the chain.
+type pdServiceBalancerNode struct {
+	ServiceClient
+	next *pdServiceBalancerNode
+}
+
+// pdServiceBalancer is a load balancer for PD service clients.
+// It is used to balance the request to all servers and manage the connections to multiple PD service nodes.
+type pdServiceBalancer struct {
```

Comment:
How about supporting multiple policies? And make the round-robin a common library?

Reply:
I am a bit confused here because it seems that there is forwarding logic in the internal `ServiceClient`, and now we have introduced an additional layer of the load balancer. Will these two things conflict with each other?

Reply:
> I am a bit confused here because it seems that there is forwarding logic in the internal `ServiceClient`, and now we have introduced an additional layer of the load balancer. Will these two things conflict with each other?

There will be no conflict. In the past, followers were randomly selected for forwarding, but now it is in turn

Reply:
> How about supporting multiple policies? And make the round-robin a common library?

I've thought about it. Since it is observed that there is now only one Round-Robin balancer in `/pkg/balancer`, I haven't implemented multiple policies yet